## 1. Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import os

## 2. Importing Data

In [2]:
# Creating a reusable path to my data

path = r'C:\Users\Vaiva\Documents\CareerFoundry\Achievement 4 07-2023 Instacart Basket Analysis'

In [3]:
# Importing customers.csv

df_cust = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

In [4]:
# Importing orders_products_flagged.pkl

df = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_flagged.pkl'))

## 3. Exploring the Data

In [5]:
df_cust.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [6]:
df.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   user_id  order_id  order_number  order_day_of_week  order_hour_of_day  ...  \
0      138   3139998            28                  6                 11  ...   
1      138   1977647            30                  6                 17  ...   
2      709    389851             2                  0                 21  ...   
3      764    652770             1                  3                 13  ...   
4      764   1813452             3                  4                 17  ...   

     price_range_loc     busiest_day    busiest_days busiest_period_of_day  \
0  Mid-range product  Regularly busy  Regularly busy           Most orders   
1  Mid-range product  Regularly busy  Regularly busy        Average orders   
2  Mid-range product     Busiest day     Busiest day        Average orders   
3  Mid-range product  Regularly busy      Least busy           Most orders   
4  Mid-range product      Least busy      Least busy        Average orders   

  max_order      loyalty_flag average_spend spending_flag  \
0        32  Regular customer      6.935811   Low Spender   
1        32  Regular customer      6.935811   Low Spender   
2         5      New customer      7.930208   Low Spender   
3         3      New customer      4.972414   Low Spender   
4         3      New customer      4.972414   Low Spender   

   days_since_prior_order_median     frequency_flag  
0                            8.0  Frequent customer  
1                            8.0  Frequent customer  
2                            8.0  Frequent customer  
3                            9.0  Frequent customer  
4                            9.0  Frequent customer  

[5 rows x 24 columns]

In [7]:
# For merging, the key 'user_id' has to be same data type, so we are checking

df.dtypes

product_id                          int64
product_name                       object
aisle_id                            int64
department_id                       int64
prices                            float64
user_id                             int64
order_id                            int64
order_number                        int64
order_day_of_week                   int64
order_hour_of_day                   int64
days_since_prior_order            float64
add_to_cart_order                   int64
reordered                           int64
ind                              category
price_range_loc                    object
busiest_day                        object
busiest_days                       object
busiest_period_of_day              object
max_order                           int64
loyalty_flag                       object
average_spend                     float64
spending_flag                      object
days_since_prior_order_median     float64
frequency_flag                    

In [8]:
# For merging, the key 'user_id' has to be same data type, so we are checking

df_cust.dtypes

user_id          int64
First Name      object
Surnam          object
Gender          object
STATE           object
Age              int64
date_joined     object
n_dependants     int64
fam_status      object
income           int64
dtype: object

In [9]:
df_cust.shape

(206209, 10)

In [10]:
df.shape

(32404859, 24)

In [11]:
df_cust.describe().astype(int)

user_id     Age  n_dependants  income
count   206209  206209        206209  206209
mean    103105      49             1   94632
std      59527      18             1   42473
min          1      18             0   25903
25%      51553      33             0   59874
50%     103105      49             1   93547
75%     154657      66             3  124244
max     206209      81             3  593901

In [12]:
df.describe().astype(int)

product_id  aisle_id  department_id    prices   user_id  order_id  \
count    32404859  32404859       32404859  32404859  32404859  32404859   
mean        25598        71              9         7    102937   1710745   
std         14083        38              6         4     59466    987298   
min             1         1              1         1         1         2   
25%         13544        31              4         4     51422    855947   
50%         25302        83              9         7    102616   1711049   
75%         37947       107             16        11    154389   2565499   
max         49688       134             21        25    206209   3421083   

       order_number  order_day_of_week  order_hour_of_day  \
count      32404859           32404859           32404859   
mean             17                  2                 13   
std              17                  2                  4   
min               1                  0                  0   
25%               5                  1                 10   
50%              11                  3                 13   
75%              24                  5                 16   
max              99                  6                 23   

       days_since_prior_order  add_to_cart_order  reordered  max_order  \
count                30328763           32404859   32404859   32404859   
mean                       11                  8          0         33   
std                         8                  7          0         25   
min                         0                  1          0          1   
25%                         5                  3          0         13   
50%                         8                  6          1         26   
75%                        15                 11          1         47   
max                        30                145          1         99   

       average_spend  days_since_prior_order_median  
count       32404859                       32404854  
mean               7                             10  
std                0                              7  
min                1                              0  
25%                7                              6  
50%                7                              8  
75%                8                             13  
max               23                             30

## 4. Cleaning Data

### Renaming columns

In [13]:
# Renaming First Name column

df_cust.rename(columns = {'First Name' : 'name'}, inplace = True)

In [14]:
# Renaming Surnam column

df_cust.rename(columns = {'Surnam' : 'surname'}, inplace = True)

In [15]:
# Renaming Gender column

df_cust.rename(columns = {'Gender' : 'gender'}, inplace = True)

In [16]:
# Renaming STATE column

df_cust.rename(columns = {'STATE' : 'state'}, inplace = True)

In [17]:
# Renaming Age column

df_cust.rename(columns = {'Age' : 'age'}, inplace = True)

In [18]:
# Checking if renaming columns worked

df_cust.head()

user_id      name   surname  gender       state  age date_joined  \
0    26711   Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890  Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803   Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935  Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797       Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

### Finding and addressing mixed data types

In [19]:
# Checking for mixed-data type

for col in df_cust.columns.tolist():
  weird = (df_cust[[col]].applymap(type) != df_cust[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_cust[weird]) > 0:
    print (col)

name


#### The 'name' column is mixed-type because there are some NaN values that are not in string type. The following lines explore this further

### Finding and addressing missing values

In [20]:
# Checking for missing values in df_cust

df_cust.isna().sum()

user_id             0
name            11259
surname             0
gender              0
state               0
age                 0
date_joined         0
n_dependants        0
fam_status          0
income              0
dtype: int64

#### Comment

There are 11259 rows with user names missing, however without contacting the stakeholders, we cannot impute them. We will leave them missing as this should not influence our analysis.

### Finding and addressing duplicates

In [21]:
# Finding duplicates in the df_cust

df_dups = df_cust[df_cust.duplicated()]

In [22]:
df_dups

Empty DataFrame
Columns: [user_id, name, surname, gender, state, age, date_joined, n_dependants, fam_status, income]
Index: []

#### No duplicates found

## 5. Merging Dataframes

In [23]:
# Due to memory issues, we are altering the data types in our dataframes in order to reduce the size of the files

# First checking the memory usage for the df dataframe

df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 24 columns):
 #   Column                         Dtype   
---  ------                         -----   
 0   product_id                     int64   
 1   product_name                   object  
 2   aisle_id                       int64   
 3   department_id                  int64   
 4   prices                         float64 
 5   user_id                        int64   
 6   order_id                       int64   
 7   order_number                   int64   
 8   order_day_of_week              int64   
 9   order_hour_of_day              int64   
 10  days_since_prior_order         float64 
 11  add_to_cart_order              int64   
 12  reordered                      int64   
 13  ind                            category
 14  price_range_loc                object  
 15  busiest_day                    object  
 16  busiest_days                   object  
 17  busiest_period_of_day    

In [24]:
# Change types for a few columns

df['order_id']=df['order_id'].astype('int32')
df['user_id'] = df['user_id'].astype('int32')
df['order_number']=df['order_number'].astype('int8')
df['order_day_of_week']=df['order_day_of_week'].astype('int8')
df['order_hour_of_day']=df['order_hour_of_day'].astype('int8')
df['days_since_prior_order']=df['days_since_prior_order'].astype('float16')

In [25]:
df.info()

# Merging customers.csv with orders_products_flagged.pkl

#df_merged = df.merge(df_cust, on = 'user_id', indicator = True)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 24 columns):
 #   Column                         Dtype   
---  ------                         -----   
 0   product_id                     int64   
 1   product_name                   object  
 2   aisle_id                       int64   
 3   department_id                  int64   
 4   prices                         float64 
 5   user_id                        int32   
 6   order_id                       int32   
 7   order_number                   int8    
 8   order_day_of_week              int8    
 9   order_hour_of_day              int8    
 10  days_since_prior_order         float16 
 11  add_to_cart_order              int64   
 12  reordered                      int64   
 13  ind                            category
 14  price_range_loc                object  
 15  busiest_day                    object  
 16  busiest_days                   object  
 17  busiest_period_of_day    

In [26]:
# 1 GB reduced. Tryng another batch

df['product_id'] =df['product_id'].astype('int32')
df['reordered']=df['reordered'].astype('int8')
df['add_to_cart_order']=df['add_to_cart_order'].astype('int32')
df['aisle_id'] = df['aisle_id'].astype('int8')
df['department_id'] = df['department_id'].astype('int8')
df['prices'] = df['prices'].astype('float16')

In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 24 columns):
 #   Column                         Dtype   
---  ------                         -----   
 0   product_id                     int32   
 1   product_name                   object  
 2   aisle_id                       int8    
 3   department_id                  int8    
 4   prices                         float16 
 5   user_id                        int32   
 6   order_id                       int32   
 7   order_number                   int8    
 8   order_day_of_week              int8    
 9   order_hour_of_day              int8    
 10  days_since_prior_order         float16 
 11  add_to_cart_order              int32   
 12  reordered                      int8    
 13  ind                            category
 14  price_range_loc                object  
 15  busiest_day                    object  
 16  busiest_days                   object  
 17  busiest_period_of_day    

In [28]:
# Checking the memory usage for the customer df

df_cust.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   name          194950 non-null  object
 2   surname       206209 non-null  object
 3   gender        206209 non-null  object
 4   state         206209 non-null  object
 5   age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [29]:
# Changing types on the customer dataframe

df_cust['user_id'] = df_cust['user_id'].astype('int32')
df_cust['age'] = df_cust['age'].astype('int8')
df_cust['n_dependants'] = df_cust['n_dependants'].astype('int8')

In [30]:
df_cust.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int32 
 1   name          194950 non-null  object
 2   surname       206209 non-null  object
 3   gender        206209 non-null  object
 4   state         206209 non-null  object
 5   age           206209 non-null  int8  
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int8  
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int32(1), int64(1), int8(2), object(6)
memory usage: 12.2+ MB


In [31]:
df_cust.head()

user_id      name   surname  gender       state  age date_joined  \
0    26711   Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890  Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803   Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935  Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797       Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [32]:
# Attempting a merge

df_merged = df.merge(df_cust, on = ['user_id'], indicator = True)

In [33]:
df_merged.head()

product_id                product_name  aisle_id  department_id     prices  \
0           1  Chocolate Sandwich Cookies        61             19   5.800781   
1           1  Chocolate Sandwich Cookies        61             19   5.800781   
2         907        Premium Sliced Bacon       106             12  20.000000   
3         907        Premium Sliced Bacon       106             12  20.000000   
4        1000                    Apricots        18             10  12.898438   

   user_id  order_id  order_number  order_day_of_week  order_hour_of_day  ...  \
0      138   3139998            28                  6                 11  ...   
1      138   1977647            30                  6                 17  ...   
2      138   3160996             1                  5                 13  ...   
3      138   2254091            10                  5                 14  ...   
4      138    505689             9                  6                 12  ...   

      name  surname  gender      state age date_joined n_dependants  \
0  Charles      Cox    Male  Minnesota  81    8/1/2019            1   
1  Charles      Cox    Male  Minnesota  81    8/1/2019            1   
2  Charles      Cox    Male  Minnesota  81    8/1/2019            1   
3  Charles      Cox    Male  Minnesota  81    8/1/2019            1   
4  Charles      Cox    Male  Minnesota  81    8/1/2019            1   

  fam_status  income _merge  
0    married   49620   both  
1    married   49620   both  
2    married   49620   both  
3    married   49620   both  
4    married   49620   both  

[5 rows x 34 columns]

In [34]:
df_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 34 columns):
 #   Column                         Dtype   
---  ------                         -----   
 0   product_id                     int32   
 1   product_name                   object  
 2   aisle_id                       int8    
 3   department_id                  int8    
 4   prices                         float16 
 5   user_id                        int32   
 6   order_id                       int32   
 7   order_number                   int8    
 8   order_day_of_week              int8    
 9   order_hour_of_day              int8    
 10  days_since_prior_order         float16 
 11  add_to_cart_order              int32   
 12  reordered                      int8    
 13  ind                            category
 14  price_range_loc                object  
 15  busiest_day                    object  
 16  busiest_days                   object  
 17  busiest_period_of_day    

In [35]:
# Checking value counts of the merge

df_merged['_merge'].value_counts()

both          32404859
left_only            0
right_only           0
Name: _merge, dtype: int64

In [36]:
# Exporting orders_products_all.pkl

df_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_all.pkl'))